services/QuillLMS/db/big_query/README.md

```diff
@@ -0,0 +1,68 @@
+
+The `views` folder contains SQL commands for creating ["Materialized Views"](https://cloud.google.com/bigquery/docs/materialized-views-intro). We are using materialized views to create tables in BigQuery that make common reporting queries more efficient by one or more methods. From BigQuery's docs:
+```
+The following use cases highlight the value of materialized views. Materialized views can improve query performance if you frequently require the following:
+
+- Pre-aggregate data. Aggregation of streaming data.
+- Pre-filter data. Run queries that only read a particular subset of the table.
+- Pre-join data. Query joins, especially between large and small tables.
+- Recluster data. Run queries that would benefit from a clustering scheme that differs from the base tables.
+```
+
+Important to know:
+- BigQuery automatically updates materialized views as new data comes in from the underlying tables
+- There are some [restrictions](https://cloud.google.com/bigquery/docs/materialized-views-create#query_limitations) on what the materialized view can contain, e.g. you can't use `OUTER JOINS`.
+- There are [tips](https://cloud.google.com/bigquery/docs/materialized-views-create#which_materialized_views_to_create) on creating efficient materialized views, e.g. put the largest, most frequently changing table first in the query
+- To minimize devops work, views should be configured to self-refresh whenever possible.
+
+### How to Create a Materialized View In SQL
+```SQL
+CREATE MATERIALIZED VIEW DATASET.MATERIALIZED_VIEW_NAME AS (
+  # QUERY_EXPRESSION
+);
+```
+### How to create a new Materialized View in the LMS
+1. Add a new entry to `config/big_query_views.yml` following pattern of other entries
+2. Add .sql file referenced in config to `db/big_query/views/`. Note, this should note include the "CREATE" command, just the inner SQL.
+2. Run a command to `create/drop/refresh` the view (you need the write permission credentials to do this locally, else do this on staging or production)
+#### Available Commands
+```
+thor quill_big_query:materialized_views:create view_key   # Creates a specified Materialized View
+thor quill_big_query:materialized_views:drop view_key     # Drops a specified Materialized View
+thor quill_big_query:materialized_views:refresh view_key  # Refreshes (drops and creates) a specified Materialized View
+```
+### Migrating to new View Version
+This is something we haven't done yet, so this policy may change, but the easiest way to change/edit a view is to create a new view with a different key, put that in place, point the code to it, and drop the old view, remove the config.
```

Comment:
No strong feelings on this one, and given that we haven't had to do this yet we don't have to have a strong opinion right now, but is there a reason not to do these migrations the same way we do them for Postgres?  That is: deploy your migration first, run migration, then deploy updated code in a secondary deploy?

Reply:
I think we might saying the same thing (I wrote this quickly, so the wording may not be as clear). What I'm saying is that instead of editing an existing view (which would break existing code during the middle of `deploy your migration first, run migration (existing code breaks), then deploy updated code (new code works)`), you should create a new view, run the migration to put it in place, then deploy the code to point to the new view. That method has no downtime.